File: soya/src/pro_42860.ts
function solution(name) {
    let answer = 0;
    let names = name.split('');
    let lchecker = 0;
    let rchecker = 0;
    
    function changeTopdown(val){
        const a = 'A'.charCodeAt();
        const z = 'Z'.charCodeAt();
        const v = val.charCodeAt();
        answer+= Math.min( v-a, z-v+1);
    }
    let totalNeedNum = 0; //'A'가 아닌 갯수
    //알파벳별 변환
    for(let i = 0 ; i < names.length ; i++){
        //알파벳 변환하기 
        if(names[i] !=='A'){
            changeTopdown(names[i]);
            totalNeedNum++;
        }
    }
   if(names[0] !=='A'){
        names[0] = 'A';
        totalNeedNum--; // 0번 인덱스는 A갯수에서 뺀다. 시작위치므로 찾을필요가 없음 
    }
    
    let minMovedCount = names.length-1;
    
    // 좌, 우로 갔을때 최적 이동수 구하기 
    function greedy(curIdx,moveCnt){
        if(moveCnt >= minMovedCount) return;
        if(totalNeedNum === 0){
            minMovedCount = Math.min(minMovedCount,moveCnt)
            return;
        }

        let idx = curIdx;
        for(let cnt = 1 ; cnt < names.length ; cnt++){
            let rnextIdx = idx+cnt < names.length ? idx+cnt : idx+cnt-names.length;
            let lnextIdx = idx-cnt < 0 ? names.length-(cnt-idx) : idx-cnt;
            if(names[rnextIdx] ==='A' && names[lnextIdx] ==='A') continue;
            if(names[rnextIdx] !== 'A'){   
                let saved = names[rnextIdx];
                names[rnextIdx] = 'A'; // 덮어쓰기
                totalNeedNum--;
                greedy(rnextIdx,moveCnt+cnt);
                names[rnextIdx] = saved;
                totalNeedNum++;
            }
            if(names[lnextIdx]!=='A'){
                let saved = names[lnextIdx];
                names[lnextIdx] = 'A'; //덮어쓰기
                totalNeedNum--;
                greedy(lnextIdx,moveCnt+cnt);
                names[lnextIdx] = saved;
                totalNeedNum++;
            }
        }  
    }

    greedy(0,0);
    answer += minMovedCount    
    
    return answer;
}

File: java/JaeHyeon/pro_42860.java
class Solution {
    public int solution(String name) {
        int answer = 0;
        int length = name.length();

        int move = length - 1; // 좌우 움직임 수를 체크

        for (int i = 0; i < name.length(); i++) {
            answer += Math.min(name.charAt(i) - 'A', 'Z' - name.charAt(i) + 1);

            int index = i + 1;
            // 연속되는 A 갯수 확인
            while (index < length && name.charAt(index) == 'A') {
                index++;
            }

            // 순서대로 가는 것과, 뒤로 돌아가는 것 중 이동수가 적은 것을 선택
            // 그냥 처음 move만큼 가는 것과, 처음으로 돌아가서 커서를 왼쪽으로 이동하여
            // A 뒤에 있는 내용을 탐색하는 경우를 계산하기 위함
            move = Math.min(move, i * 2 + length - index);

            // 뒤로 돌아가는 것, 뒤로 갔다가 다시 앞으로 오기 때문에 *2,
            // +i를 해주는 이유는 일단 index가 A인 구간을 만나서 계산이 들어가기 때문
            move = Math.min(move, (length - index) * 2 + i);
        }

        return answer + move;
    }
}